File: go-backend-interview-prep/11_system_design_in_go/01_rate_limiter_service_mini/main.py
"""Rate limiter service -- Python equivalent using token bucket."""

import json
import threading
import time
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.request import urlopen, Request
from urllib.error import URLError


# --- Token Bucket ---

class Bucket:
    def __init__(self, max_tokens, refill_rate):
        self.tokens = max_tokens
        self.max_tokens = max_tokens
        self.refill_rate = refill_rate  # tokens per second
        self.last_refill = time.time()

    def allow(self):
        now = time.time()
        elapsed = now - self.last_refill
        self.tokens += elapsed * self.refill_rate
        if self.tokens > self.max_tokens:
            self.tokens = self.max_tokens
        self.last_refill = now

        if self.tokens >= 1:
            self.tokens -= 1
            return True
        return False


# --- Rate Limiter ---

class RateLimiter:
    def __init__(self, refill_rate, max_tokens):
        self.lock = threading.Lock()
        self.buckets = {}
        self.max_tokens = max_tokens
        self.refill_rate = refill_rate

    def allow(self, key):
        with self.lock:
            if key not in self.buckets:
                self.buckets[key] = Bucket(self.max_tokens, self.refill_rate)
            return self.buckets[key].allow()


# --- Handler with rate limit ---

limiter = RateLimiter(refill_rate=2, max_tokens=5)


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/ping":
            key = self.client_address[0]
            if not limiter.allow(key):
                self.send_response(429)
                self.send_header("Content-Type", "application/json")
                self.send_header("Retry-After", "1")
                self.end_headers()
                self.wfile.write(json.dumps({"error": "rate limit exceeded"}).encode())
                return

            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"message": "pong"}).encode())
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, format, *args):
        pass  # suppress default logs


# --- Demo ---

def run_demo():
    time.sleep(0.5)
    for i in range(1, 9):
        try:
            req = Request("http://localhost:9001/ping")
            resp = urlopen(req, timeout=2)
            body = json.loads(resp.read())
            print(f"req {i}: status={resp.status} body={body}")
        except URLError as e:
            if hasattr(e, "code"):
                print(f"req {i}: status={e.code} body=rate limit exceeded")
            else:
                print(f"req {i}: error: {e}")

    print("\n--- waiting 2s for token refill ---")
    time.sleep(2)

    try:
        resp = urlopen("http://localhost:9001/ping", timeout=2)
        body = json.loads(resp.read())
        print(f"req after wait: status={resp.status} body={body}")
    except URLError as e:
        if hasattr(e, "code"):
            print(f"req after wait: status={e.code}")
        else:
            print(f"req after wait: error: {e}")

    print("\ndemo done")
    import os
    os._exit(0)


if __name__ == "__main__":
    print("rate limiter server on :9001 (2 tokens/sec, burst 5)")
    t = threading.Thread(target=run_demo, daemon=True)
    t.start()
    server = HTTPServer(("", 9001), Handler)
    server.serve_forever()

File: go-backend-interview-prep/11_system_design_in_go/01_rate_limiter_service_mini/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"sync"
	"time"
)

// --- Token Bucket ---

// Bucket tracks tokens for a single client.
type Bucket struct {
	tokens     float64
	maxTokens  float64
	refillRate float64 // tokens per second
	lastRefill time.Time
}

// Allow checks if a request is allowed and consumes a token.
func (b *Bucket) Allow() bool {
	now := time.Now()
	elapsed := now.Sub(b.lastRefill).Seconds()
	b.tokens += elapsed * b.refillRate
	if b.tokens > b.maxTokens {
		b.tokens = b.maxTokens
	}
	b.lastRefill = now

	if b.tokens >= 1 {
		b.tokens--
		return true
	}
	return false
}

// --- Rate Limiter ---

// RateLimiter manages per-client buckets.
type RateLimiter struct {
	mu         sync.Mutex
	buckets    map[string]*Bucket
	maxTokens  float64
	refillRate float64
}

// NewRateLimiter creates a limiter with given rate and burst.
func NewRateLimiter(refillRate, maxTokens float64) *RateLimiter {
	return &RateLimiter{
		buckets:    make(map[string]*Bucket),
		maxTokens:  maxTokens,
		refillRate: refillRate,
	}
}

// Allow checks if the given client key is within rate limit.
func (rl *RateLimiter) Allow(key string) bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	b, ok := rl.buckets[key]
	if !ok {
		b = &Bucket{
			tokens:     rl.maxTokens,
			maxTokens:  rl.maxTokens,
			refillRate: rl.refillRate,
			lastRefill: time.Now(),
		}
		rl.buckets[key] = b
	}
	return b.Allow()
}

// --- Middleware ---

func rateLimitMiddleware(rl *RateLimiter, next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		key := r.RemoteAddr // per-client by IP
		if !rl.Allow(key) {
			w.Header().Set("Retry-After", "1")
			w.WriteHeader(http.StatusTooManyRequests)
			json.NewEncoder(w).Encode(map[string]string{
				"error": "rate limit exceeded",
			})
			return
		}
		next(w, r)
	}
}

// --- Handler ---

func pingHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{
		"message": "pong",
	})
}

// --- Demo ---

func main() {
	// 2 tokens/sec, burst of 5
	limiter := NewRateLimiter(2, 5)

	mux := http.NewServeMux()
	mux.HandleFunc("/ping", rateLimitMiddleware(limiter, pingHandler))

	// Demo: simulate requests in a goroutine
	go func() {
		time.Sleep(500 * time.Millisecond)
		client := &http.Client{Timeout: 2 * time.Second}
		for i := 1; i <= 8; i++ {
			resp, err := client.Get("http://localhost:9001/ping")
			if err != nil {
				fmt.Printf("req %d: error: %v\n", i, err)
				continue
			}
			var body map[string]string
			json.NewDecoder(resp.Body).Decode(&body)
			resp.Body.Close()
			fmt.Printf("req %d: status=%d body=%v\n", i, resp.StatusCode, body)
		}

		// Wait for refill, then try again
		fmt.Println("\n--- waiting 2s for token refill ---")
		time.Sleep(2 * time.Second)

		resp, err := client.Get("http://localhost:9001/ping")
		if err != nil {
			fmt.Printf("req after wait: error: %v\n", err)
		} else {
			var body map[string]string
			json.NewDecoder(resp.Body).Decode(&body)
			resp.Body.Close()
			fmt.Printf("req after wait: status=%d body=%v\n", resp.StatusCode, body)
		}

		log.Println("demo done -- shutting down")
		go func() { time.Sleep(100 * time.Millisecond); log.Fatal("exit") }()
	}()

	fmt.Println("rate limiter server on :9001 (2 tokens/sec, burst 5)")
	log.Fatal(http.ListenAndServe(":9001", mux))
}
